crates/ironrdp-graphics/src/rdp6/rle.rs

```diff
@@ -159,6 +166,185 @@ pub fn decompress_8bpp_plane(
+        let mut result = (next as i16 - prev as i16) as u8;
+
+        // bit magic from 3.1.9.2.1 of [MS-RDPEGDI].
+        if result < 128 {
+            result <<= 1;
+        } else {
+            result = (255u8.wrapping_sub(result) << 1).wrapping_add(1);
+        }
+
+        result
+    }
+}
+
+impl<I: Iterator<Item = u8>> Iterator for RleEncoderScanlineIterator<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Some((idx, mut next)) = self.inner.next() else {
+            return None;
+        };
+
+        let prev = std::mem::replace(&mut self.prev_scanline[idx % self.width], next);
+        if idx >= self.width {
+            next = self.delta_value(prev, next);
+        }
+
+        Some(next)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+#[derive(Debug)]
+struct RlePlaneEncoder {
+    width: usize,
+    height: usize,
+}
+
+impl RlePlaneEncoder {
+    pub fn new(width: usize, height: usize) -> Self {
+        Self { width, height }
+    }
+
+    pub fn encode(&self, mut src: impl Iterator<Item = u8>, dst: &mut WriteCursor<'_>) -> Result<usize, RleError> {
+        let mut written = 0;
+
+        for _ in 0..self.height {
+            written += self.encode_scanline((&mut src).take(self.width), dst)?;
+        }
+
+        Ok(written)
+    }
+
+    fn encode_scanline(&self, mut src: impl Iterator<Item = u8>, dst: &mut WriteCursor<'_>) -> Result<usize, RleError> {
+        let mut written = 0;
+
+        let Some(first) = src.next() else {
+            return Err(RleError::NotEnoughBytes);
+        };
+
+        let mut raw = vec![first];
+        let mut seq = (first, 0);
+
+        for byte in src {
+            let (last, count) = seq;
+
+            seq = if byte == last {
+                (byte, count + 1)
+            } else {
+                match count {
+                    3.. => {
+                        written += self.encode_segment(&raw, count, dst);
+                        raw.clear();
+                    }
+                    2 => raw.extend_from_slice(&[last, last]),
+                    1 => raw.push(last),
+                    _ => {}
+                }
+
+                raw.push(byte);
+
+                (byte, 0)
+            }
+        }
+
+        let (last, mut count) = seq;
+        if count < 3 {
+            raw.extend(vec![last; count].into_iter());
+            count = 0;
+        }
+
+        written += self.encode_segment(&raw, count, dst);
+
+        Ok(written)
+    }
+
+    fn encode_segment(&self, mut raw: &[u8], run: usize, dst: &mut WriteCursor<'_>) -> usize {
+        let mut extra_bytes = 0;
+
+        while raw.len() > 15 {
+            extra_bytes += self.encode_segment(&raw[0..15], 0, dst);
+            raw = &raw[15..];
+        }
+
+        let control = ((raw.len() as u8) << 4) + std::cmp::min(run, 15) as u8;
+        dst.write_u8(control);
```

Comment:
We need to ensure that the destination cursor has enough bytes here via `ensure_size`, because any `write_*` or `read_*` methods on cursor will panic if there is not enough space

Reply:
done